File: 0025.reverse-nodes-in-k-group/25.reverse-nodes-in-k-group.py
#
# @lc app=leetcode id=25 lang=python3
#
# [25] Reverse Nodes in k-Group
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if k <= 1:
            return head
        root, pre, left = head, None, head
        while left and left.next:
            i, right = 1, left.next
            while right.next and i < k-1:
                right = right.next
                i+=1
            if i == k-1:
                tmp = right.next
                self.reverseList(left, right)
                if root == head:
                    root = right
                if pre:
                    pre.next = right
                pre = left
                left = tmp
            else:
                left = None
        return root
        

    def reverseList(self, head: ListNode, tail: ListNode) -> ListNode:
        if head==None or head.next == None or tail == None:
            return head
        cur = head
        while cur and cur != tail:
            tmp = head.next.next
            head.next.next = cur
            cur = head.next
            head.next = tmp
        return tail

# @lc code=end

File: 0025.reverse-nodes-in-k-group/25.reverse-nodes-in-k-group.go
package main

/*
 * @lc app=leetcode id=25 lang=golang
 *
 * [25] Reverse Nodes in k-Group
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	var newHead *ListNode
	var first *ListNode
	for cur := head; cur != nil; {
		i, nodes := 0, make([]*ListNode, k)
		for ; i < k && cur != nil; i++ {
			nodes[i] = cur
			cur = cur.Next
		}
		if i < k {
			if first != nil {
				first.Next = nodes[0]
			}
			break
		}
		if first != nil {
			first.Next = nodes[k-1]
		}
		if newHead == nil {
			newHead = nodes[k-1]
		}
		for j := k - 1; j > 0; j-- {
			nodes[j].Next = nodes[j-1]
		}
		if cur == nil {
			nodes[0].Next = nil
		} else {
			first = nodes[0]
		}
	}
	return newHead
}

// @lc code=end
